torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -637,41 +645,63 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
   return func_prefix + "_" + std::to_string(value);
 }
 
-Stmt* GPUMetaVarRewriter::mutate(const For* v) {
-  // Recurse first.
-  Stmt* body = v->body()->accept_mutator(this);
+bool GPUMetaVarRewriter::isFullRank() {
+  {
+    auto& extents = cuda_analysis_->gpu_block_extents();
+    for (int i = 0; i < 3; ++i) {
+      if (!exprEquals(current_block_rank_[i], extents[i])) {
+        return false;
+      }
+    }
+  }
 
+  {
+    auto& extents = cuda_analysis_->gpu_thread_extents();
+    for (int i = 0; i < 3; ++i) {
+      if (!exprEquals(current_thread_rank_[i], extents[i])) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+Stmt* GPUMetaVarRewriter::mutate(const For* v) {
+  Stmt* body = v->body();
+  const Expr* old_rank = nullptr;
   const LoopOptions& loop_options = v->loop_options();
   if (loop_options.is_gpu_block_index()) {
-    auto& extents = cuda_analysis_->gpu_block_extents();
     int gpu_block_index = loop_options.gpu_block_index();
     if (gpu_block_index >= 3) {
       throw std::runtime_error("support only 3D gpu_block_index");
     }
+    old_rank = current_block_rank_[gpu_block_index];
+
+    // Extents must be positive, assume >= 1.
+    if (old_rank->isConstant() && immediateEquals(old_rank, 1)) {
+      current_block_rank_[gpu_block_index] = v->stop();
```

Comment:
Minor: I am a bit troubled by the term "rank" here. Until you see it somewhere in Cuda guide, most places use the term "rank" to refer to the rank of a tensor. I would prefer to pick a different name here. But it is up to you.

Reply:
Yeah, happy to - any suggestions?

Reply:
I want to avoid "dimension", "extent" and "size" because they're overloaded as well. I landed on "reach" ??